### DISCUSSION TITLE
Uniformly Distributed Points

### DISCUSSION THREAD

**Discussion by 19kdc3** [marcomusy/vedo#716] [General]
Hi Marco,

I'm wondering if you have any suggestions to uniformly distribute a set number of points throughout a mesh. The mesh will be either 2D or 3D, and in the case of 3D, points should also be distributed throughout the body (i.e. not only the surface). These points are later used to probe the original mesh.

Currently I've tried decimating the original mesh setting "n" to the number of points/probes I want, but I run into 2 issues:
1. The result doesn't always have "n" number of points
2. The result points aren't uniformly distributed

![image](https://user-images.githubusercontent.com/73371848/197572052-d682b999-e2f4-449e-a241-155f40e61fbe.png)

Thanks for your help!

**Comment by marcomusy** [comment]
Hi @kman750 what about this:

```python
from vedo import *

msh = Polygon(nsides=5)

pts = utils.pack_spheres(msh, radius=0.1)

# optionally add some noise:
jitter = np.random.randn(len(pts),3)/1000
jitter[:,2] = 0
pts += jitter

pts = Points(pts)
pts.cut_with_line(msh.points()) # needs an ordered set of points

show(msh, pts, axes=1)
```
![Screenshot from 2022-10-24 20-56-59](https://user-images.githubusercontent.com/32848391/197604150-27441c3e-d261-48a6-a7ff-4e5cade44f4b.png)

I would not use `decimate`.
Also you can extend easily this to 3D (you can use `cut_with_mesh()` or `some_mesh.inside_points(my_points)` to test the points).

**Reply by 19kdc3** [reply]
Is there a way to specify the number of points? (e.x. I want to specify that 20 points be distributed) 

The reason I had used decimate was that it allowed me to specify "n" number of points, though it doesn't always seem to follow that specification.

**Reply by marcomusy** [reply]
I think the problem has no general solution, you can only roughly compute the right radius but then you need to find the solution iteratively..
(you can use `msh.area()` and in 3D `mhs.volume()` to help for that).

**Reply by 19kdc3** [reply]
If the points don't need to be **perfectly** uniformly distributed, just more equally to the point that no obvious tight clustering is occurring, are there any tricks or tools you may be able to point me in the direction of?

**Reply by marcomusy** [reply]
The above already provides what you are asking for!
Or I may have misunderstood what you mean.

**Reply by 19kdc3** [reply]
Sorry, let me be more clear. Your above solution works with regards to uniformly distributing points (points just need to be evenly distributed with no obvious clustering), but it's missing the ability to specify the number of points.

**Reply by marcomusy** [reply]
You cannot have both things, I'm not aware of any algorithm that can give you the  _exact_ nr of points, unless, as I said you first make an estimate based on the area/volume and then refine it by trying different values of `radius`.

**Reply by 19kdc3** [reply]
Thank you very much for your help and suggestions Marco! I really appreciate it.

**Reply by marcomusy** [reply]
You're welcome!

**Reply by RyanConway91** [reply]
Perhaps an iterative solution? Generate grids of different spacings and then query the number of points in the grid and repeat until have the desired number of points? Not sure what solver you would use to run the optimization though.

**Reply by marcomusy** [reply]
I think any non-gradient based optimizer would work - there are 2 examples here:
https://github.com/marcomusy/vedo/blob/master/examples/other/iminuit1.py (must choose "simplex")
https://github.com/marcomusy/vedo/blob/master/examples/other/nevergrad_opt.py

..or simple trial and error :)
